### ISSUE TITLE
New go.mod files causes issues 

### ISSUE THREAD

**Issue by morphar** [google/gnostic#283]
After renaming apps directory to cmd, it's no longer possible to `go install` any of the apps without cloning the repo.
There is no issues when cloning the repo and running go install from within the wanted cmd directory (e.g.: cmd/protoc-gen-openapi).
It's strictly an issue concerning non-cloned use of the modules.

(go version go1.17 darwin/amd64)

### How to reproduce
From a project with or without gnostic in go.mod.
When I run: `go install github.com/google/gnostic/cmd/protoc-gen-openapi`
I get:
```text
no required module provides package github.com/google/gnostic/cmd/protoc-gen-openapi; to add it:
	go get github.com/google/gnostic/cmd/protoc-gen-openapi
```
When I run: `go get github.com/google/gnostic/cmd/protoc-gen-openapi`
I get:
```text
go get: github.com/google/gnostic/cmd/protoc-gen-openapi@v0.0.0-20220107061744-94bcf11351a5 requires
	github.com/google/gnostic@v0.0.0: reading github.com/google/gnostic/go.mod at revision v0.0.0: unknown revision v0.0.0
```

When I run `go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest` outside of any repos
I get:
```text
go install: github.com/google/gnostic/cmd/protoc-gen-openapi@latest (in github.com/google/gnostic/cmd/protoc-gen-openapi@v0.0.0-20220107061744-94bcf11351a5):
	The go.mod file for the module providing named packages contains one or
	more replace directives. It must not contain directives that would cause
	it to be interpreted differently than if it were the main module.
```

So it seems there is no longer any way to get the apps, without cloning the gnostic repo first.

@timburks
This seems like a hard issue to fix as it needs to be published, before we know if it's fixed or not.
A couple of ideas for a fix:
1. Remove the go.mod files for each of the modules in cmd
2. Remove the v0.0.0 and `replace` directive in all the go.mod files in cmd

What are the benefits of having a go.mod for each command?
I may be missing something, but as far as I understand, it makes sense for tighter dependency control of each command.
But is this needed or even wanted in this case?
Doesn't it actually make more sense to have 1 go.mod file in the root, since everything is more or less interdependent, which makes it even more necessary to _**not**_ have 2 commands installed from 2 different versions of the repo.

**Comment by morphar** [comment]
@timburks I have created a pull request (#284), in case we want to go back to 1 go.mod file.

**Comment by timburks** [comment]
@morphar Thanks, I agree with reverting the multi-module go.mod files. The alternative (referring to specific versions from the cmd/go.mod files) seems difficult to maintain. The tighter dependencies are helpful for protecting from security vulnerabilities, but not required.

I'll merge #284 now.

**Comment by morphar** [comment]
@timburks better security makes sense. I do agree though, that maintaining self-referencing go.mod versions would be very difficult.
Is it possible to update the current release and / or delete the v0.6.0?

**Comment by timburks** [comment]
I can't reuse the tag (that causes problems downstream) but just pushed a v0.6.1 tag with your changes.

**Comment by morphar** [comment]
👍 Of course - makes sense :)

**Comment by shenqidebaozi** [comment]
@timburks @morphar Due to my negligence, I found that there was a `replace` instruction in the previous code submission. The replace instruction cannot appear in the CMD type module, so we only need to remove the `replace` instruction. I also replied to this problem in the previous PR comment

**Comment by shenqidebaozi** [comment]
https://github.com/google/gnostic/pull/279#discussion_r776554948

**Comment by shenqidebaozi** [comment]
We only need to remove the `replace` instruction, not go.mod

**Comment by timburks** [comment]
@shenqidebaozi No worries, I added the replace statements to the `go.mod` files in `cmd` hoping to avoid explicitly specifying gnostic versions in their dependencies. From your perspective, what do we lose by having a single `go.mod` at the root of the repo?

**Comment by shenqidebaozi** [comment]
@timburks For `go install`, the fact that the `replace` instruction cannot appear in the CMD type program is the reason why it cannot be installed correctly, so we only need to remove the replace instruction.
https://github.com/golang/go/issues/44840

**Comment by morphar** [comment]
I'm not sure, that would actually fix the issue.
The gnostic@v0.0.0 also seems to create some issues.
Since the go.mods are "self-referencing", it would be difficult to maintain and update.
E.g.: Whenever a command needs to use something updated in some other parts of the repo, we would need to make a release, then update the command to use that release and then make a new release, with the updated dependency.

I might be totally wrong or missing something here, it's just what it looks like from my perspective.

**Comment by tonybase** [comment]
This error occurred while I was installing:
```
go install: github.com/google/gnostic/cmd/protoc-gen-openapi@latest

(in github.com/google/gnostic/cmd/protoc-gen-openapi@v0.0.0-20220107061744-94bcf11351a5):
	The go.mod file for the module providing named packages contains one or
	more replace directives. It must not contain directives that would cause
	it to be interpreted differently than if it were the main module.
make: *** [init] Error 1
```

**Comment by shenqidebaozi** [comment]
@morphar I think the `proto-gen-openapi` plug-in is an independent application. It should have its own `go.mod` file. If it does not have its own `go.mod` file, the user needs to install `gnostic` first, but I just want to use this separate plug-in, not `gnostic`

**Comment by shenqidebaozi** [comment]
@morphar I think the `proto-gen-openapi` plug-in is an independent application. It should have its own `go.mod` file. If it does not have its own `go.mod` file, the user needs to install `gnostic` first, but I just want to use this separate plug-in, not `gnostic`

**Comment by morphar** [comment]
@shenqidebaozi you can install `protoc-gen-openapi` in single module mode with:
`go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest`
right?

**Comment by shenqidebaozi** [comment]
Many similar projects are managed in a multi module manner.
Therefore, I think we only need to delete the `replace` instruction

**Comment by morphar** [comment]
@tonybase I'm getting that as well, which is weird, since we just merged a fix.
It seems like it using the release instead of the tag, when using @latest.
This works:
`go install github.com/google/gnostic/cmd/protoc-gen-openapi@v0.6.1`

**Comment by shenqidebaozi** [comment]
> you can install `protoc-gen-openapi` in single module mode with:
> `go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest`
> right?

Yes, if it has a separate gomod file, I just need `go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest`

**Comment by shenqidebaozi** [comment]
> @shenqidebaozi you can install in single module mode with: right?`protoc-gen-openapi``go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest`

If it doesn't have a separate `go.mod `file, I can only pull the `gnostic` first

**Comment by morphar** [comment]
@shenqidebaozi I think we are misunderstanding each other.
The current v0.6.1 tag has the single module code, which will work with:
`go install github.com/google/gnostic/cmd/protoc-gen-openapi@v0.6.1`
This will install the `protoc-gen-openapi` command.
I do understand, that this will download the entire `gnostic` repo.
The thing is: `protoc-gen-openapi` is **_not_** self-contained. It depends on other parts of gnostic, so it will be downloaded even if `protoc-gen-openapi` has it's own go.mod, since that file will have a dependency on `gnostic`.

So in my perspective, it seems there is no gains and only extra work in maintenance and release.

**Comment by morphar** [comment]
@timburks it seems we need and actual release to make this work again:
`github.com/google/gnostic/cmd/protoc-gen-openapi@latest`
It ignores the tag and uses the release.

**Comment by shenqidebaozi** [comment]
@morphar Yes, I see what you mean. I'm sorry for the current situation. It seems that we can only release one version to fix the problem that the `@latest` version cannot be specified.

**Comment by morphar** [comment]
@shenqidebaozi No worries, I'm just trying to understand if I'm missing something and get to the best solution.
So just to make sure: Do we agree that the 1 module solution is the best? At least for now?

**Comment by shenqidebaozi** [comment]
@morphar Single module and multi module are two completely different management methods. Multi module does have the problem of increasing workload, just like you said "So in my perspective, it seems there is no gains and only extra work in maintenance and release."

**Comment by shenqidebaozi** [comment]
@morphar If we don't have much time to manage complex multiple modules, I think a single module is more suitable for 'gnostic'`

**Comment by morphar** [comment]
@shenqidebaozi 👍 I do not oppose the multi module idea at all. It just seems like there are better fits than Gnostic.
E.g. the original example you had (`grpc/grpc-go/cmd/protoc-gen-go-grpc/`), this module only has `google.golang.org/protobuf` as a requirement, so it does not "self-reference" back to the root module.
In this case you would not end up with a need to do double-releases and you would get the benefit of not needing the root module.

In the case of `gnostic` there is no actual gain like that.
It seems that we agree :)

**Comment by shenqidebaozi** [comment]
@morphar Yes, it may be that our project adopts multi module management, so the actual scenario of `gnostic` is not considered when submitting pr. I'm very sorry for causing some trouble, but renaming the `app` to `cmd` is also a very correct decision. I hope to provide better code contribution to `gnostic` in the future.

**Comment by morphar** [comment]
@shenqidebaozi No worries 👍 I totally agree on the `app` -> `cmd` change. I have been thinking the same.
It's unfortunate that it causes so many problems, but things like that **will** happen.
This was a really hard thing to test for, so it slipped through.
We were 3 people looking at the code and agreeing that it seemed correct, so there wasn't much more we could have done.

**Comment by shenqidebaozi** [comment]
@morphar I think `go` dependence on management is always unsatisfactory. Our previous project was also troubled.

**Comment by shenqidebaozi** [comment]
However, it has also accumulated experience for us. For me, this is a bumpy but quite pleasant open source contribution

**Comment by morphar** [comment]
@shenqidebaozi I actually haven't had any issues with it. I loved it when it came and replaced `godep`.
When I start a project, I try to think structure and constraints, before I start, so that I know how my Docker Compose setup should work, how go modules are built and how the git structure is set up - I believe that safes me from a lot of headaches down the road, as I usually don't run into to many problems.

The biggest issue I have, is the replace for local dev, where I also have to build container with the replace code, but it seems like that is also solved in Go 1.18.

**Comment by morphar** [comment]
@shenqidebaozi everything seems to have been fixed now :) 👍
